src/utils/src/InstrumentedAllocators.c

```diff
@@ -0,0 +1,148 @@
+#include "Include_i.h"
+
+volatile SIZE_T gInstrumentedAllocatorsTotalAllocationSize = 0;
+
+memAlloc gInstrumentedAllocatorsStoredMemAlloc;
+memAlignAlloc gInstrumentedAllocatorsStoredMemAlignAlloc;
+memCalloc gInstrumentedAllocatorsStoredMemCalloc;
+memFree gInstrumentedAllocatorsStoredMemFree;
+memRealloc gInstrumentedAllocatorsStoredMemRealloc;
+
+STATUS setInstrumentedAllocators()
+{
+    // Store the existing function pointers
+    gInstrumentedAllocatorsStoredMemAlloc = globalMemAlloc;
+    gInstrumentedAllocatorsStoredMemAlignAlloc = globalMemAlignAlloc;
+    gInstrumentedAllocatorsStoredMemCalloc = globalMemCalloc;
+    gInstrumentedAllocatorsStoredMemFree = globalMemFree;
+    gInstrumentedAllocatorsStoredMemRealloc = globalMemRealloc;
+
+    // Overwrite with the instrumented ones
+    globalMemAlloc = instrumentedAllocatorsMemAlloc;
+    globalMemAlignAlloc = instrumentedAllocatorsMemAlignAlloc;
+    globalMemCalloc = instrumentedAllocatorsMemCalloc;
+    globalMemFree = instrumentedAllocatorsMemFree;
+    globalMemRealloc = instrumentedAllocatorsMemRealloc;
+
+    return STATUS_SUCCESS;
+}
+
+STATUS resetInstrumentedAllocators()
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    SIZE_T totalRemainingSize = ATOMIC_LOAD(&gInstrumentedAllocatorsTotalAllocationSize);
+
+    // Reset the global allocators with the stored ones
+    globalMemAlloc = gInstrumentedAllocatorsStoredMemAlloc;
+    globalMemAlignAlloc = gInstrumentedAllocatorsStoredMemAlignAlloc;
+    globalMemCalloc = gInstrumentedAllocatorsStoredMemCalloc;
+    globalMemFree = gInstrumentedAllocatorsStoredMemFree;
+    globalMemRealloc = gInstrumentedAllocatorsStoredMemRealloc;
+
+    // Check the final total value
+    CHK_WARN(totalRemainingSize == 0, STATUS_MEMORY_NOT_FREED,
+            "Possible memory leak of size %" PRIu64, totalRemainingSize);
+
+CleanUp:
+
+    CHK_LOG_ERR_NV(retStatus);
+    return retStatus;
+}
+
+SIZE_T getInstrumentedTotalAllocationSize()
+{
+    SIZE_T totalRemainingSize = ATOMIC_LOAD(&gInstrumentedAllocatorsTotalAllocationSize);
+    return totalRemainingSize;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Internal functionality
+////////////////////////////////////////////////////////////////////////////////
+
+PVOID instrumentedAllocatorsMemAlloc(SIZE_T size)
+{
+    DLOGS("Instrumented mem alloc %" PRIu64 " bytes", (UINT64) size);
+    PSIZE_T pAlloc = (PSIZE_T) gInstrumentedAllocatorsStoredMemAlloc(size + SIZEOF(SIZE_T));
+
+    if (pAlloc == NULL) {
+        return NULL;
+    }
+
+    // Store the allocation size - should be aligned
+    *pAlloc = size;
+
+    // Add to the total book keeping
+    ATOMIC_ADD(&gInstrumentedAllocatorsTotalAllocationSize, size);
+
+    return pAlloc + 1;
+}
+
+PVOID instrumentedAllocatorsMemAlignAlloc(SIZE_T size, SIZE_T alignment)
+{
+    // Doing the same as alloc
+    DLOGS("Instrumented mem alignalloc %" PRIu64 " bytes", (UINT64) size);
```

Comment:
Why arent we supporting this right now?

Reply:
The main reason is that in the KVS Client we have a flag that sets up the content store as the allocator which will be tightly packed by default. In this case, the align alloc will be meaningless. Also, we don't use this anywhere in the codebase. I think the best would be to simply remove this API at some stage.